app/org/sagebionetworks/bridge/play/controllers/ConsentController.java

```diff
@@ -134,30 +143,36 @@ Result changeSharingScope(SharingScope sharingScope, String message) {
         final UserSession session = getAuthenticatedAndConsentedSession();
         final Study study = studyService.getStudy(session.getStudyIdentifier());
         
+        updateSharingStatusAndSession(study, session, sharingScope);
+        return okResult(message);
+    }
+
+    private void updateSharingStatusAndSession(Study study, UserSession session, SharingScope sharingScope) {
         optionsService.setEnum(study, session.getHealthCode(), SHARING_SCOPE, sharingScope);
         
-        StudyParticipant participant = new StudyParticipant.Builder()
-                .copyOf(session.getParticipant())
+        StudyParticipant participant = new StudyParticipant.Builder().copyOf(session.getParticipant())
                 .withSharingScope(sharingScope).build();
         session.setParticipant(participant);
         updateSession(session);
-        return okResult(message);
     }
-
+    
     private Result giveConsentForVersion(int version, SubpopulationGuid subpopGuid) throws Exception {
         final UserSession session = getAuthenticatedSession();
         final Study study = studyService.getStudy(session.getStudyIdentifier());
 
-        final ConsentSignature consent = parseJson(request(), ConsentSignature.class);
+        final ConsentSignature consentSignature = parseJson(request(), ConsentSignature.class);
         final SharingOption sharing = SharingOption.fromJson(requestToJSON(request()), version);
 
-        consentService.consentToResearch(study, subpopGuid, session, consent,
+        ConsentStatus status = session.getConsentStatuses().get(subpopGuid);
+        if (status == null) {
+            throw new EntityNotFoundException(Subpopulation.class);
+        }
+        
+        consentService.consentToResearch(study, subpopGuid, session.getParticipant(), consentSignature,
                 sharing.getSharingScope(), true);
         
-        StudyParticipant participant = new StudyParticipant.Builder().copyOf(session.getParticipant())
-                .withSharingScope(sharing.getSharingScope()).build();
-        session.setParticipant(participant);
-        updateSession(session);
+        UserSession updatedSession = authenticationService.updateSession(study, getCriteriaContext(session));
+        updateSession(updatedSession);
```

Comment:
Replaced updateSession() with one method that does all this (gets new session, sets it everywhere), and used that everywhere instead.